File: src/spark_e2e/runner.py
"""YAML test runner — runs test scenarios defined in YAML files sequentially.

Usage:
    spark-e2e run                    # runs tests/*.yaml
    spark-e2e run path/to/test.yaml  # runs a single file
"""

from __future__ import annotations

import json
import os
import time
from pathlib import Path
from typing import Any

import yaml


# ── Types ──────────────────────────────────────────────────


def _load_yaml(path: str) -> dict[str, Any]:
    with open(path) as f:
        return yaml.safe_load(f)


def _find_test_files(pattern: str | None = None) -> list[str]:
    """Find YAML test files. If pattern is given, use it directly or scan dir."""
    if pattern:
        resolved = os.path.abspath(pattern)
        if os.path.isfile(resolved):
            return [resolved]
        if os.path.isdir(resolved):
            return _scan_dir(resolved)
        print(f"No test files found at: {pattern}")
        return []

    # Default: scan tests/
    tests_dir = os.path.join(os.getcwd(), "tests")
    if not os.path.isdir(tests_dir):
        return []
    return _scan_dir(tests_dir)


def _scan_dir(directory: str) -> list[str]:
    files: list[str] = []
    try:
        for entry in sorted(os.listdir(directory)):
            full = os.path.join(directory, entry)
            if os.path.isdir(full):
                files.extend(_scan_dir(full))
            elif os.path.isfile(full) and os.path.splitext(entry)[1] in (".yaml", ".yml"):
                files.append(full)
    except OSError:
        pass
    return files


# ── Step execution ─────────────────────────────────────────


def _run_steps(
    scenarios: list[dict],
    browser: Any,
    provider: Any,
    cfg: Any,
    base_url: str,
    output_dir: str,
) -> list[dict]:
    """Execute all scenarios and return results."""
    from spark_e2e.vlm.openai_compat import extract_json
    from spark_e2e import prompts as prompts_mod

    scenario_results: list[dict] = []

    for scenario in scenarios:
        name = scenario.get("name", "(unnamed)")
        if not name:
            scenario_results.append({"name": "(unnamed)", "pass": False, "steps": []})
            continue

        print(f"\n  Scenario: {name}")
        step_results: list[dict] = []

        for step in scenario.get("steps", []):
            started = time.time()
            result = _exec_step(step, browser, provider, cfg, base_url, output_dir, extract_json, prompts_mod)
            result["durationMs"] = int((time.time() - started) * 1000)

            icon = "✓" if result["pass"] else "✗"
            detail = str(result.get("detail", ""))[:80]
            print(f"    {icon} {result['type']}: {detail}")
            step_results.append(result)

        all_pass = all(r["pass"] for r in step_results)
        scenario_results.append({"name": name, "pass": all_pass, "steps": step_results})

    return scenario_results


def _exec_step(
    step: dict,
    browser: Any,
    provider: Any,
    cfg: Any,
    base_url: str,
    output_dir: str,
    extract_json: Any,
    prompts_mod: Any,
) -> dict:
    """Execute a single step and return its result dict."""

    def _capture():
        png = browser.capture_screenshot(reload=False)
        return browser.to_data_url(png)

    def _locate(target: str) -> dict:
        data_url = _capture()
        prompt = prompts_mod.get_locate_prompt(target)
        raw = provider.chat(prompt, data_url, None, cfg.vlm.thinking_budget)
        try:
            return extract_json(raw)
        except Exception:
            return {"found": False, "reasoning": "VLM parse error"}

    # ── navigate ──
    if "navigate" in step:
        url = step["navigate"]
        if not url.startswith("http"):
            url = base_url.rstrip("/") + "/" + url.lstrip("/")
        browser.navigate(url)
        return {"step": step, "pass": True, "type": "navigate", "detail": url}

    # ── wait ──
    if "wait" in step:
        browser.wait_for_timeout(int(step["wait"] * 1000))
        return {"step": step, "pass": True, "type": "wait", "detail": f"{step['wait']}s"}

    # ── snapshot ──
    if "snapshot" in step:
        png = browser.capture_screenshot(reload=False)
        os.makedirs(output_dir, exist_ok=True)
        filename = "".join(c if c.isalnum() or c in "_-" else "_" for c in step["snapshot"]) + ".png"
        out_path = os.path.join(output_dir, filename)
        with open(out_path, "wb") as f:
            f.write(png)
        return {"step": step, "pass": True, "type": "snapshot", "detail": out_path}

    # ── click ──
    if "click" in step:
        loc = _locate(step["click"])
        if not loc.get("found"):
            return {"step": step, "pass": False, "type": "click", "detail": f"Not found: {loc.get('reasoning', '?')}"}
        browser.click_at(float(loc["x"]), float(loc["y"]))
        return {"step": step, "pass": True, "type": "click", "detail": f"({loc['x']}, {loc['y']})"}

    # ── type ──
    if "type" in step:
        target = step["type"]["into"]
        text = step["type"]["text"]
        loc = _locate(target)
        if not loc.get("found"):
            return {"step": step, "pass": False, "type": "type", "detail": f"Target not found: {loc.get('reasoning', '?')}"}
        browser.click_at(float(loc["x"]), float(loc["y"]))
        browser.wait_for_timeout(150)
        browser.clear_and_type(text)
        return {"step": step, "pass": True, "type": "type", "detail": f'"{text}" into ({loc["x"]}, {loc["y"]})'}

    # ── hover ──
    if "hover" in step:
        loc = _locate(step["hover"])
        if not loc.get("found"):
            return {"step": step, "pass": False, "type": "hover", "detail": f"Not found: {loc.get('reasoning', '?')}"}
        browser.hover_at(float(loc["x"]), float(loc["y"]))
        return {"step": step, "pass": True, "type": "hover", "detail": f"({loc['x']}, {loc['y']})"}

    # ── test ──
    if "test" in step:
        data_url = _capture()
        prompt = prompts_mod.get_test_prompt(step["test"], cfg.prompts.strictness)
        raw = provider.chat(prompt, data_url, None, cfg.vlm.thinking_budget)
        try:
            result = extract_json(raw)
            passed = result.get("pass") is True
            detail = result.get("summary", "passed" if passed else "failed")
        except Exception:
            passed = False
            detail = str(raw)[:200]
        return {"step": step, "pass": passed, "type": "test", "detail": str(detail)}

    # ── assert ──
    if "assert" in step:
        data_url = _capture()
        prompt_lines = [
            "You are a visual E2E test verifier. Determine whether this assertion is TRUE or FALSE.",
            f"ASSERTION: {step['assert']}",
            "",
            prompts_mod.get_assert_prompt(cfg.prompts.strictness),
            "",
            'Respond ONLY with JSON: {"pass": true|false, "confidence": "high"|"medium"|"low", "observation": "...", "reasoning": "..."}',
        ]
        raw = provider.chat("\n".join(prompt_lines), data_url, None, cfg.vlm.thinking_budget)
        try:
            result = extract_json(raw)
            passed = result.get("pass") is True
            detail = result.get("reasoning", result.get("summary", "passed" if passed else "failed"))
        except Exception:
            passed = False
            detail = str(raw)[:200]
        return {"step": step, "pass": passed, "type": "assert", "detail": str(detail)}

    return {"step": step, "pass": False, "type": "unknown", "detail": "Unknown step type"}


# ── Public API ─────────────────────────────────────────────


def run_tests(pattern: str | None = None) -> list[dict]:
    """Run YAML test files and return a list of file reports."""
    from spark_e2e.config import load
    from spark_e2e.browser import get_backend
    from spark_e2e.vlm import get_provider

    files = _find_test_files(pattern)

    if not files:
        print("No test files found. Create tests/*.yaml files or specify a path.")
        print("Example: spark-e2e run tests/login.test.yaml")
        return []

    reports: list[dict] = []
    for filepath in files:
        print(f"\n📄 {filepath}")
        try:
            suite = _load_yaml(filepath)
            scenarios = suite.get("scenarios", [])
            if not isinstance(scenarios, list) or not scenarios:
                print("  ✗ ERROR: no 'scenarios' list found")
                reports.append({"file": filepath, "pass": False, "scenarios": [], "durationMs": 0})
                continue

            cfg = load()
            backend = get_backend(cfg.browser.backend)
            provider = get_provider(cfg.vlm.provider)
            base_url = suite.get("config", {}).get("url", cfg.browser.url)

            # Apply viewport if specified in YAML
            vp = suite.get("config", {}).get("viewport", {})
            if vp.get("width") and vp.get("height"):
                backend._ensure_browser()
                backend._page.set_viewport_size({"width": vp["width"], "height": vp["height"]})

            basename = os.path.splitext(os.path.basename(filepath))[0]
            output_dir = os.path.join(os.getcwd(), ".spark-e2e", "runs", basename)

            started = time.time()
            scenario_results = _run_steps(scenarios, backend, provider, cfg, base_url, output_dir)
            duration_ms = int((time.time() - started) * 1000)

            all_pass = all(s["pass"] for s in scenario_results)
            status = "✓ PASS" if all_pass else "✗ FAIL"
            print(f"  {status} ({duration_ms / 1000:.1f}s)")

            reports.append({
                "file": filepath,
                "pass": all_pass,
                "scenarios": scenario_results,
                "durationMs": duration_ms,
            })

            backend.close()

        except Exception as e:
            print(f"  ✗ ERROR: {e}")
            reports.append({"file": filepath, "pass": False, "scenarios": [], "durationMs": 0})

    return reports

File: src/runner.ts
/**
 * YAML test runner — declares scenarios as YAML, runs them sequentially.
 *
 * Usage:
 *   spark-e2e run                    # runs tests/*.yaml
 *   spark-e2e run path/to/test.yaml  # runs a single file
 */

import { existsSync, readFileSync, writeFileSync, mkdirSync, readdirSync, statSync } from "node:fs";
import { resolve, join, basename, extname } from "node:path";
import * as yaml from "js-yaml";
import { getConfig, interpolateEnvVars } from "./config.js";
import { PlaywrightBrowser } from "./browser/playwright.js";
import { getProvider } from "./vlm/index.js";
import { getTestPrompt, getAssertPrompt, getLocatePrompt } from "./prompts.js";
import { extractJson } from "./vlm/openai-compat.js";

// ── Types ──────────────────────────────────────────────────

interface TestSuite {
  name?: string;
  config?: {
    url?: string;
    viewport?: { width?: number; height?: number };
  };
  scenarios: Scenario[];
}

interface Scenario {
  name: string;
  steps: Step[];
}

type Step =
  | { navigate: string }
  | { click: string }
  | { type: { text: string; into: string } }
  | { hover: string }
  | { test: string }
  | { assert: string }
  | { wait: number }
  | { snapshot: string };

interface StepResult {
  step: Record<string, unknown>;
  pass: boolean;
  type: string;
  detail: string;
  durationMs: number;
}

interface ScenarioResult {
  name: string;
  pass: boolean;
  steps: StepResult[];
}

interface RunReport {
  file: string;
  pass: boolean;
  scenarios: ScenarioResult[];
  durationMs: number;
}

// ── Step execution ────────────────────────────────────────

async function runStep(
  step: Step,
  browser: PlaywrightBrowser,
  provider: ReturnType<typeof getProvider>,
  cfg: ReturnType<typeof getConfig>,
  baseUrl: string,
  outputDir: string,
): Promise<StepResult> {
  const started = Date.now();

  // ── navigate ──
  if ("navigate" in step) {
    const url = step.navigate.startsWith("http") ? step.navigate : new URL(step.navigate, baseUrl).href;
    await browser.navigate(url);
    return { step: step as Record<string, unknown>, pass: true, type: "navigate", detail: url, durationMs: Date.now() - started };
  }

  // ── wait ──
  if ("wait" in step) {
    await browser.waitForTimeout(step.wait * 1000);
    return { step: step as Record<string, unknown>, pass: true, type: "wait", detail: `${step.wait}s`, durationMs: Date.now() - started };
  }

  // ── snapshot ──
  if ("snapshot" in step) {
    const buf = await browser.captureScreenshot();
    const filename = `${step.snapshot.replace(/[^a-zA-Z0-9_-]/g, "_")}.png`;
    const outPath = join(outputDir, filename);
    mkdirSync(outputDir, { recursive: true });
    writeFileSync(outPath, Buffer.isBuffer(buf) ? buf : Buffer.from(buf as ArrayBuffer));
    return { step: step as Record<string, unknown>, pass: true, type: "snapshot", detail: outPath, durationMs: Date.now() - started };
  }

  // ── VLM-based steps (click, type, hover, test, assert) ──

  const { dataUrl } = await captureCurrent(browser, cfg);

  // ── click ──
  if ("click" in step) {
    if (step.click.startsWith("@")) {
      const coords = resolveDomRef(step.click);
      if (!coords) return { step: step as Record<string, unknown>, pass: false, type: "click", detail: `Dom ref not found: ${step.click}`, durationMs: Date.now() - started };
      await browser.clickAt(coords.x, coords.y);
      return { step: step as Record<string, unknown>, pass: true, type: "click", detail: `${step.click} (${coords.x}, ${coords.y}) [dom-ref]`, durationMs: Date.now() - started };
    }
    const locate = await locateElement(provider, cfg, dataUrl, step.click);
    if (!locate.found) return { step: step as Record<string, unknown>, pass: false, type: "click", detail: `Not found: ${locate.reasoning}`, durationMs: Date.now() - started };
    await browser.clickAt(locate.x, locate.y);
    return { step: step as Record<string, unknown>, pass: true, type: "click", detail: `(${locate.x}, ${locate.y})`, durationMs: Date.now() - started };
  }

  // ── type ──
  if ("type" in step) {
    if (step.type.into.startsWith("@")) {
      const coords = resolveDomRef(step.type.into);
      if (!coords) return { step: step as Record<string, unknown>, pass: false, type: "type", detail: `Dom ref not found: ${step.type.into}`, durationMs: Date.now() - started };
      await browser.clickAt(coords.x, coords.y);
      await browser.waitForTimeout(150);
      await browser.clearAndType(step.type.text);
      return { step: step as Record<string, unknown>, pass: true, type: "type", detail: `[masked] into ${step.type.into} [dom-ref]`, durationMs: Date.now() - started };
    }
    const locate = await locateElement(provider, cfg, dataUrl, step.type.into);
    if (!locate.found) return { step: step as Record<string, unknown>, pass: false, type: "type", detail: `Target not found: ${locate.reasoning}`, durationMs: Date.now() - started };
    await browser.clickAt(locate.x, locate.y);
    await browser.waitForTimeout(150);
    await browser.clearAndType(step.type.text);
    return { step: step as Record<string, unknown>, pass: true, type: "type", detail: `[masked] into (${locate.x}, ${locate.y})`, durationMs: Date.now() - started };
  }

  // ── hover ──
  if ("hover" in step) {
    if (step.hover.startsWith("@")) {
      const coords = resolveDomRef(step.hover);
      if (!coords) return { step: step as Record<string, unknown>, pass: false, type: "hover", detail: `Dom ref not found: ${step.hover}`, durationMs: Date.now() - started };
      await browser.hoverAt(coords.x, coords.y);
      return { step: step as Record<string, unknown>, pass: true, type: "hover", detail: `${step.hover} (${coords.x}, ${coords.y}) [dom-ref]`, durationMs: Date.now() - started };
    }
    const locate = await locateElement(provider, cfg, dataUrl, step.hover);
    if (!locate.found) return { step: step as Record<string, unknown>, pass: false, type: "hover", detail: `Not found: ${locate.reasoning}`, durationMs: Date.now() - started };
    await browser.hoverAt(locate.x, locate.y);
    return { step: step as Record<string, unknown>, pass: true, type: "hover", detail: `(${locate.x}, ${locate.y})`, durationMs: Date.now() - started };
  }

  // ── test ──
  if ("test" in step) {
    const prompt = getTestPrompt(step.test, cfg.prompts.strictness);
    const raw = await provider.chat(prompt, dataUrl, undefined, cfg.vlm.thinkingBudget);
    let result: Record<string, unknown> = {};
    try { result = extractJson(raw) as Record<string, unknown>; } catch { result = { pass: false, summary: raw.slice(0, 200) }; }
    const pass = result.pass === true;
    return { step: step as Record<string, unknown>, pass, type: "test", detail: (result.summary as string) || (pass ? "passed" : "failed"), durationMs: Date.now() - started };
  }

  // ── assert ──
  if ("assert" in step) {
    const prompt = [
      "You are a visual E2E test verifier. Determine whether this assertion is TRUE or FALSE.",
      `ASSERTION: ${step.assert}`,
      "",
      getAssertPrompt(cfg.prompts.strictness),
      "",
      'Respond ONLY with JSON: {"pass": true|false, "confidence": "high"|"medium"|"low", "observation": "...", "reasoning": "..."}',
    ].join("\n");
    const raw = await provider.chat(prompt, dataUrl, undefined, cfg.vlm.thinkingBudget);
    let result: Record<string, unknown> = {};
    try { result = extractJson(raw) as Record<string, unknown>; } catch { result = { pass: false, reasoning: raw.slice(0, 200) }; }
    const pass = result.pass === true;
    return { step: step as Record<string, unknown>, pass, type: "assert", detail: (result.reasoning as string) || (result.summary as string) || (pass ? "passed" : "failed"), durationMs: Date.now() - started };
  }

  // Should not reach here for valid YAML
  return { step: step as Record<string, unknown>, pass: false, type: "unknown", detail: "Unknown step type", durationMs: Date.now() - started };
}

// ── Helpers ────────────────────────────────────────────────

async function captureCurrent(browser: PlaywrightBrowser, cfg: ReturnType<typeof getConfig>): Promise<{ dataUrl: string }> {
  const buf = await browser.captureScreenshot({ reload: false, maskSelectors: cfg.security.maskSelectors });
  const raw = Buffer.isBuffer(buf) ? buf : Buffer.from(buf as ArrayBuffer);
  return { dataUrl: browser.toDataUrl(raw) };
}

function resolveDomRef(ref: string): { x: number; y: number } | null {
  const domStatePath = resolve(process.cwd(), ".spark", "plugin", "e2e", "dom-state.json");
  if (!existsSync(domStatePath)) return null;
  try {
    const state = JSON.parse(readFileSync(domStatePath, "utf-8"));
    const el = state.layout?.find((e: { ref: string }) => e.ref === ref);
    if (!el) return null;
    return { x: el.center.x, y: el.center.y };
  } catch {
    return null;
  }
}

interface LocateResult {
  found: boolean;
  x: number;
  y: number;
  reasoning?: string;
}

async function locateElement(
  provider: ReturnType<typeof getProvider>,
  cfg: ReturnType<typeof getConfig>,
  dataUrl: string,
  target: string,
): Promise<LocateResult> {
  const prompt = getLocatePrompt(target);
  const raw = await provider.chat(prompt, dataUrl, undefined, cfg.vlm.thinkingBudget);
  try {
    const result = extractJson(raw) as Record<string, unknown>;
    return {
      found: result.found === true,
      x: (result.x as number) || 0,
      y: (result.y as number) || 0,
      reasoning: result.reasoning as string,
    };
  } catch {
    return { found: false, x: 0, y: 0, reasoning: "VLM returned unparseable response" };
  }
}

// ── Run one file ──────────────────────────────────────────

async function runFile(yamlPath: string): Promise<RunReport> {
  const started = Date.now();
  const raw = readFileSync(yamlPath, "utf-8");
  const suite = interpolateEnvVars(yaml.load(raw)) as unknown as TestSuite;

  if (!suite.scenarios || !Array.isArray(suite.scenarios)) {
    throw new Error(`Invalid test file: ${yamlPath} — missing "scenarios" list`);
  }

  const cfg = getConfig();
  const browser = new PlaywrightBrowser();
  const provider = getProvider(cfg.vlm.provider);
  const baseUrl = suite.config?.url || cfg.browser.url || "";

  // Apply viewport from YAML config if present
  if (suite.config?.viewport?.width && suite.config?.viewport?.height) {
    const page = await browser.ensurePage();
    await page.setViewportSize({
      width: suite.config.viewport.width,
      height: suite.config.viewport.height,
    });
  }

  const outputDir = resolve(process.cwd(), ".spark", "plugin", "e2e", "runs", basename(yamlPath, ".yaml"));

  const scenarioResults: ScenarioResult[] = [];

  for (const scenario of suite.scenarios) {
    if (!scenario.name) {
      scenarioResults.push({ name: "(unnamed)", pass: false, steps: [] });
      continue;
    }

    console.error(`\n  Scenario: ${scenario.name}`);
    const stepResults: StepResult[] = [];

    for (const step of scenario.steps) {
      const result = await runStep(step, browser, provider, cfg, baseUrl, outputDir);
      const icon = result.pass ? "✓" : "✗";
      console.error(`    ${icon} ${result.type}: ${result.detail.slice(0, 80)}`);
      stepResults.push(result);
      // Stop scenario on first failure? Or continue? Let's continue by default.
    }

    const allPassed = stepResults.every((r) => r.pass);
    scenarioResults.push({ name: scenario.name, pass: allPassed, steps: stepResults });
  }

  await browser.close();

  return {
    file: yamlPath,
    pass: scenarioResults.every((s) => s.pass),
    scenarios: scenarioResults,
    durationMs: Date.now() - started,
  };
}

// ── File discovery ────────────────────────────────────────

function findTestFiles(pattern?: string): string[] {
  // If a specific file is given, use it directly
  if (pattern) {
    const resolved = resolve(pattern);
    if (existsSync(resolved) && statSync(resolved).isFile()) return [resolved];
    // If it's a directory, scan it
    if (existsSync(resolved) && statSync(resolved).isDirectory()) {
      return scanDir(resolved);
    }
    console.error(`No test files found at: ${pattern}`);
    return [];
  }

  // Default: scan tests/ directory
  const testsDir = resolve(process.cwd(), "tests");
  if (!existsSync(testsDir)) return [];
  return scanDir(testsDir);
}

function scanDir(dir: string): string[] {
  const files: string[] = [];
  try {
    for (const entry of readdirSync(dir, { withFileTypes: true })) {
      const fullPath = join(dir, entry.name);
      if (entry.isDirectory()) {
        files.push(...scanDir(fullPath));
      } else if (entry.isFile() && [".yaml", ".yml"].includes(extname(entry.name))) {
        files.push(fullPath);
      }
    }
  } catch {
    // Permission errors, etc. — skip
  }
  return files.sort();
}

// ── Public API ─────────────────────────────────────────────

export async function runTests(pattern?: string): Promise<RunReport[]> {
  const files = findTestFiles(pattern);

  if (files.length === 0) {
    console.error("No test files found. Create tests/*.yaml files or specify a path.");
    console.error("Example: spark-e2e run tests/login.test.yaml");
    return [];
  }

  const reports: RunReport[] = [];
  for (const file of files) {
    console.error(`\n📄 ${file}`);
    try {
      const report = await runFile(file);
      reports.push(report);
      const status = report.pass ? "✓ PASS" : "✗ FAIL";
      console.error(`  ${status} (${(report.durationMs / 1000).toFixed(1)}s)`);
    } catch (err) {
      console.error(`  ✗ ERROR: ${(err as Error).message}`);
      reports.push({
        file,
        pass: false,
        scenarios: [],
        durationMs: 0,
      });
    }
  }

  return reports;
}
